### DISCUSSION TITLE
Ability to run system commands and capture the output in Typst

### DISCUSSION THREAD

**Discussion by PEWOLO** [typst/typst#3278] [Q&A] [answered]
Hello beautiful community members!

I was wondering if there is any possibility to run system commands and capture the output in Typst.
Any idea would be super helpful!

**Comment by andreasKroepelin** [comment, marked as answer]
Only via some tricks. The general strategy is:

1. Insert `#metadata` in your Typst document, containing the information you want to send to the external program.
2. From outside Typst, in a shell script or a dedicated program, run the `typst query` commandline and obtain your previously provided metadata in JSON form.
3. Process this data however you want to in that external program.
4. Make the external program output a JSON, TOML, YAML, XML, CSV, or CBOR (recommended for binary data) file containing the results of your processing.
5. Again in your Typst document, read that file using the respective data loading function at the same place where you've also put the `#metadata`.
6. Transform the read results into your desired output.

Now you kinda run Typst and your external program taking turns. Hope this helps!

**Reply by YPares** [reply]
Is there a way to filter #metadata uses based on contents of the value field?

For instance say my doc contains some:

```typst
#metadata((command: "sh", script: "... # script contents ..."))
```

Then doing the following doesn't work:

`$ typst query 'metadata.where(value: (command: "sh"))'`

because whatever you feed to `where` must _exactly_ match the value, not simply a part of it, but what I want are all the occurences of calls to `metadata` that contains a dict `value` with a `command: "sh"` field.

Second problem: `typst query` doesn't give you the position in the doc of the metadata call (which `line:col` it starts at and which `line:col` it ends at), that's not convenient to replace that call later on once my command has been evaluated. Cf https://github.com/typst/typst/issues/7008

**Comment by Dherse** [comment]
Outside of what was suggested, it has been discussed at length and decided that it's not worth fracturing the ecosystem between the different platforms and the security issues and risks.

**Reply by astrale-sharp** [reply]
Hi beautiful being,
An attempt to automate this is being worked on here https://github.com/ntjess/showman

**Reply by PEWOLO** [reply]
Thank so much for the link, I will give it a try right away
